### PR TITLE
Add optional span property for dashboard card

### PR DESCRIPTION
**Pull request by rawagner**
Allows adding more cards in one line.
grid-span

@cloudbehl

### CODE REVIEW

File: frontend/packages/console-plugin-sdk/src/typings/dashboards.ts [hunk 1]
```diff
@@ -56,6 +57,9 @@ namespace ExtensionProperties {
 
     /** Loader for the corresponding dashboard card component. */
     loader: LazyLoader<any>;
+
+    /** The card span in grid. Property is ignored for small screen and defaults to 12. */
+    span?: gridItemSpanValueShape;
```
Comment (rawagner) [inline frontend/packages/console-plugin-sdk/src/typings/dashboards.ts]: `4 || 6 || 12`

Comment (rawagner) [reply frontend/packages/console-plugin-sdk/src/typings/dashboards.ts]: @vojtechszocs type updated

File: frontend/packages/console-plugin-sdk/src/typings/dashboards.ts [hunk 2]
```diff
@@ -56,6 +56,9 @@ namespace ExtensionProperties {
 
     /** Loader for the corresponding dashboard card component. */
     loader: LazyLoader<any>;
+
+    /** The card span in grid. Property is ignored for small screen and defaults to 12. */
```
Comment (vojtechszocs) [inline frontend/packages/console-plugin-sdk/src/typings/dashboards.ts]: Nit: `Card's vertical span in the column. Ignored for small screens, defaults to 12.`

File: frontend/public/components/dashboard/grid.tsx [hunk 3]
```diff
@@ -3,16 +3,17 @@ import { Grid, GridItem } from '@patternfly/react-core';
...
   MAIN = 'MAIN',
   LEFT = 'LEFT',
   RIGHT = 'RIGHT',
 }
 
-const mapCardsToGrid = (cards: React.ComponentType<any>[], keyPrefix: string): React.ReactNode[] =>
-  cards.map((Card, index) => (
-    <GridItem key={`${keyPrefix}-${index}`} span={12}><Card /></GridItem>
+const mapCardsToGrid = (cards: GridDashboardCard[], keyPrefix: string, ignoreCardSpan?: boolean): React.ReactNode[] =>
+  cards.map(({Card, span}, index) => (
```
Comment (vojtechszocs) [inline frontend/public/components/dashboard/grid.tsx]: Nit: use a default value `span = 12` to simplify the below expression:

```ts
span={(ignoreCardSpan && 12) || span}
```

File: frontend/public/components/dashboard/grid.tsx [hunk 4]
```diff
@@ -3,16 +3,17 @@ import { Grid, GridItem } from '@patternfly/react-core';
...
 export enum GridPosition {
   MAIN = 'MAIN',
   LEFT = 'LEFT',
   RIGHT = 'RIGHT',
 }
 
-const mapCardsToGrid = (cards: React.ComponentType<any>[], keyPrefix: string): React.ReactNode[] =>
-  cards.map((Card, index) => (
-    <GridItem key={`${keyPrefix}-${index}`} span={12}><Card /></GridItem>
+const mapCardsToGrid = (cards: GridDashboardCard[], keyPrefix: string, ignoreCardSpan?: boolean): React.ReactNode[] =>
```
Comment (vojtechszocs) [inline frontend/public/components/dashboard/grid.tsx]: Nit: it's better to be explicit with optional value fallbacks:

```diff
-ignoreCardSpan?: boolean
+ignoreCardSpan: boolean = false
```

File: frontend/public/components/dashboard/grid.tsx [hunk 5]
```diff
@@ -59,8 +60,13 @@ export const DashboardGrid: React.FC<DashboardGridProps> = ({ mainCards, leftCar
   return <div ref={containerRef}>{grid}</div>;
 };
 
+export type GridDashboardCard = {
+  span?: DashboardCardSpan;
+  Card: React.ComponentType<any>;
```
Comment (vojtechszocs) [inline frontend/public/components/dashboard/grid.tsx]: Nit: put `Card` above `span`, key/essential properties should go before optional/non-essential ones.

File: frontend/public/components/dashboards-page/dashboards.tsx [hunk 6]
```diff
@@ -7,11 +7,14 @@ import * as plugins from '../../plugins';
...
-import { DashboardGrid, GridPosition } from '../dashboard/grid';
+import { DashboardGrid, GridPosition, GridDashboardCard } from '../dashboard/grid';
 import { DashboardsCard } from '@console/plugin-sdk';
 
-const getCardsOnPosition = (cards: DashboardsCard[], position: GridPosition): React.ComponentType<any>[] =>
-  cards.filter(c => c.properties.position === position).map(c => () => <AsyncComponent loader={c.properties.loader} />);
+const getCardsOnPosition = (cards: DashboardsCard[], position: GridPosition): GridDashboardCard[] =>
+  cards.filter(c => c.properties.position === position).map(c => ({
+    span: c.properties.span,
+    Card: () => <AsyncComponent loader={c.properties.loader} />,
```
Comment (vojtechszocs) [inline frontend/public/components/dashboards-page/dashboards.tsx]: Same as above, please put `Card` above `span`.

File: frontend/packages/console-plugin-sdk/src/typings/dashboards.ts (hunk 2 above)
Comment (rawagner) [reply frontend/packages/console-plugin-sdk/src/typings/dashboards.ts]: fixed

File: frontend/public/components/dashboard/grid.tsx (hunk 4 above)
Comment (rawagner) [reply frontend/public/components/dashboard/grid.tsx]: fixed

File: frontend/public/components/dashboard/grid.tsx (hunk 3 above)
Comment (rawagner) [reply frontend/public/components/dashboard/grid.tsx]: ok, using default value. I changed the expression to `span={ignoreCardSpan ? 12 : span}` which I personally find more readable :)

File: frontend/public/components/dashboard/grid.tsx (hunk 5 above)
Comment (rawagner) [reply frontend/public/components/dashboard/grid.tsx]: fixed

File: frontend/public/components/dashboards-page/dashboards.tsx (hunk 6 above)
Comment (rawagner) [reply frontend/public/components/dashboards-page/dashboards.tsx]: fixed

**Pull request merged**